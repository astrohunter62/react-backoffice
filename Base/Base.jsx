import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router'
import { withStyles } from 'material-ui/styles'

import withRoot from './withRoot'
import BaseBranch from './BaseBranch'

const withBase = (Component) => class extends React.Component {
  constructor() {
    super()

    this.state = {
      open: false,
      cookieInfoOpen: this.getCookie(),
    }

    this.handleCookieInfoAccept = this.handleCookieInfoAccept.bind(this)
    this.handleDrawerOpen = this.handleDrawerOpen.bind(this)
    this.handleDrawerClose = this.handleDrawerClose.bind(this)
    this.redirectTo = this.redirectTo.bind(this)
    this.onClick = this.onClick.bind(this)
  }

  getCookie() {
    const value = '; ' + document.cookie
    const parts = value.split('; cookie_concent=')

    return !(
      parts.length != 2 ?
        undefined :
        parts.pop().split(';').shift()
    )
  }

  handleDrawerOpen() {
    this.setState({
      open: true
    })
  }

  handleDrawerClose() {
    this.setState({
      open: false
    })
  }

  redirectTo(link) {
    this.props.history.push(link)
  }

  onClick() {
    this.redirectTo('/')
  }

  handleCookieInfoAccept() {
    this.setState({
      cookieInfoOpen: false,
    })
  }

  render() {
    return (
      <Component
        {...this.props}
        {...this.state}
        onClick={this.onClick}
        handleDrawerOpen={this.handleDrawerOpen}
        handleDrawerClose={this.handleDrawerClose}
        onCookieInfoAccept={this.handleCookieInfoAccept}
        redirectTo={this.redirectTo}
        rightContent={this.props.rightContent}
      />
    )
  }
}

let Base = withBase(BaseBranch)
Base = withRouter(Base)

Base.propTypes = {
  title: PropTypes.string.isRequired,
  menuData: PropTypes.arrayOf(PropTypes.object).isRequired,
  rightContent: PropTypes.element,
  fixedHeader: PropTypes.bool,
}

Base.defaultProps = {
  fixedHeader: true
}

export default withRoot(Base)
