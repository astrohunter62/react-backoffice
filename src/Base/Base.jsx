import React from 'react'
import PropTypes from 'prop-types'

import Cookie from '../CookieInfo/Cookie'
import BaseBranch from './BaseBranch'

const withBase = Component => class extends React.Component {
  static propTypes = {
    title: PropTypes.string.isRequired,
    menuOpen: PropTypes.bool,
    menuData: PropTypes.arrayOf(PropTypes.object).isRequired,
    rightContent: PropTypes.node,
    isHeaderFixed: PropTypes.bool,
    hasHeader: PropTypes.bool,
    hasCookieInfo: PropTypes.bool,
    history: PropTypes.objectOf(PropTypes.any),
  }

  static defaultProps = {
    isHeaderFixed: true,
    hasHeader: true,
    hasCookieInfo: false,
    menuOpen: false,
    rightContent: null,
    history: undefined,
  }

  constructor(props) {
    super(props)

    this.state = {
      open: false,
      cookieInfoOpen: false,
    }

    if (props.hasCookieInfo && Cookie.getCookie() === undefined) {
      Cookie.setCookie(false)
    }

    this.handleCookieInfoAccept = this.handleCookieInfoAccept.bind(this)
    this.handleDrawerOpen = this.handleDrawerOpen.bind(this)
    this.handleDrawerClose = this.handleDrawerClose.bind(this)
    this.redirectTo = this.redirectTo.bind(this)
    this.onClick = this.onClick.bind(this)
  }

  componentDidMount() {
    const { hasCookieInfo, menuOpen } = this.props

    this.setState({
      cookieInfoOpen: hasCookieInfo && Cookie.getCookie() === false,
      open: menuOpen,
    })
  }

  componentWillReceiveProps({ menuOpen }) {
    this.setState({
      open: menuOpen,
    })
  }

  onClick() {
    this.redirectTo('/')
  }

  handleDrawerOpen() {
    this.setState({
      open: true,
    })
  }

  handleDrawerClose() {
    this.setState({
      open: false,
    })
  }

  redirectTo(link) {
    const { history } = this.props

    if (history) {
      history.push(link)
    }
  }

  handleCookieInfoAccept() {
    this.setState({
      cookieInfoOpen: false,
    })
  }

  render() {
    const { rightContent } = this.props

    return (
      <Component
        {...this.props}
        {...this.state}
        onClick={this.onClick}
        handleDrawerOpen={this.handleDrawerOpen}
        handleDrawerClose={this.handleDrawerClose}
        onCookieInfoAccept={this.handleCookieInfoAccept}
        redirectTo={this.redirectTo}
        rightContent={rightContent}
      />
    )
  }
}

const WithBase = withBase(BaseBranch)

export default WithBase
