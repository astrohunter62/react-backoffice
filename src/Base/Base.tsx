import React from "react";
import { MenuDataItem } from "../Menu/Menu";
import BaseBranch from "./BaseBranch";

type WithBaseProps = {
  title: string;
  menuOpen?: boolean;
  menuData: MenuDataItem[];
  rightContent?: React.ReactNode;
  isHeaderFixed?: boolean;
  hasHeader?: boolean;
  history?: {
    [key: string]: any;
  };
};

type WithBaseState = {
  open: boolean | undefined;
};

const withBase = (Component: any) =>
  class WithBase extends React.Component<WithBaseProps, WithBaseState> {
    constructor(props: WithBaseProps) {
      super(props);
      this.state = {
        open: false,
      };
      this.handleDrawerOpen = this.handleDrawerOpen.bind(this);
      this.handleDrawerClose = this.handleDrawerClose.bind(this);
      this.redirectTo = this.redirectTo.bind(this);
      this.onClick = this.onClick.bind(this);
    }
    componentDidMount() {
      const { menuOpen } = this.props;
      this.setState({
        open: menuOpen,
      });
    }
    UNSAFE_componentWillReceiveProps({ menuOpen }: WithBaseProps) {
      this.setState({
        open: menuOpen,
      });
    }
    onClick() {
      this.redirectTo("/");
    }
    handleDrawerOpen() {
      this.setState({
        open: true,
      });
    }
    handleDrawerClose() {
      this.setState({
        open: false,
      });
    }
    redirectTo(link: string) {
      const { history } = this.props;
      if (history) {
        history.push(link);
      }
    }

    render() {
      const { rightContent, menuOpen, ...rest } = this.props;

      return (
        <Component
          {...rest}
          {...this.state}
          onClick={this.onClick}
          handleDrawerOpen={this.handleDrawerOpen}
          handleDrawerClose={this.handleDrawerClose}
          redirectTo={this.redirectTo}
          rightContent={rightContent}
        />
      );
    }
  };

export default withBase(BaseBranch);
