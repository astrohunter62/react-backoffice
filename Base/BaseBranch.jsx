import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import withStyles from 'material-ui/styles/withStyles'

import Drawer from '../Drawer'
import Header from '../Header'

const drawerWidth = 280

const styles = theme => ({
  appFrame: {
    position: 'relative',
    display: 'flex',
    width: '100%',
    height: '100%',
    minHeight: '100vh',
    transition: '0.25s',
  },
  appFrameWithCookieInfo: {
    marginTop: theme.spacing.unit * 6,
    minHeight: `calc(100vh - ${theme.spacing.unit * 6}px)`,
  },
  content: {
    width: '100%',
    flexGrow: 1,
    backgroundColor: theme.palette.background.default,
    padding: theme.spacing.unit * 3,
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    height: 'calc(100% - 56px)',
    marginTop: theme.spacing.unit * 8,
    [theme.breakpoints.up('sm')]: {
      content: {
        height: 'calc(100% - 64px)',
        marginTop: 64,
      },
    },
    marginLeft: -drawerWidth,
  },
  contentShift: {
    transition: theme.transitions.create('margin', {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
    marginLeft: 0,
  },
})

const BaseBranch = ({
  open,
  title,
  menuData,
  fixedHeader,
  onClick,
  cookieInfoOpen,
  handleDrawerOpen,
  handleDrawerClose,
  redirectTo,
  rightContent,
  hasHeader,
  classes,
  children,
  ...rest
}) => (
  <div>
    {hasHeader ? (
      <Header
        title={title}
        handleDrawerOpen={handleDrawerOpen}
        onClick={onClick}
        open={open}
        fixed={fixedHeader}
        cookieInfoOpen={cookieInfoOpen}
      >
        {rightContent || null}
      </Header>
    ) : null}

    <div className={classNames(classes.appFrame, {
      [classes.appFrameWithCookieInfo]: cookieInfoOpen,
    })}
    >
      {hasHeader ? (
        <Drawer
          handleDrawerClose={handleDrawerClose}
          redirectTo={redirectTo}
          open={open}
          data={menuData}
        />
      ) : null}

      <main
        className={classNames(classes.content, {
          [classes.contentShift]: open || !hasHeader,
        })}
      >
        {React.Children.map(children, child => (
          React.cloneElement(child, {
            cookieInfoOpen,
            ...rest,
          })
        ))}
      </main>
    </div>
  </div>
)

BaseBranch.propTypes = {
  open: PropTypes.bool,
  title: PropTypes.string.isRequired,
  menuData: PropTypes.arrayOf(PropTypes.object).isRequired,
  fixedHeader: PropTypes.bool,
  cookieInfoOpen: PropTypes.bool,
  onClick: PropTypes.func.isRequired,
  handleDrawerOpen: PropTypes.func.isRequired,
  handleDrawerClose: PropTypes.func.isRequired,
  redirectTo: PropTypes.func.isRequired,
  rightContent: PropTypes.element,
  hasHeader: PropTypes.bool.isRequired,
  classes: PropTypes.objectOf(PropTypes.string).isRequired,
  children: PropTypes.node.isRequired,
}

BaseBranch.defaultProps = {
  open: false,
  fixedHeader: false,
  cookieInfoOpen: false,
  rightContent: (<Fragment />),
}

export default withStyles(styles)(BaseBranch)
