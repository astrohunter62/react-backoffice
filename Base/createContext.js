import { create, SheetsRegistry } from 'jss'
import { jssPreset, createMuiTheme } from 'material-ui/styles'
import { blueGrey, orange } from 'material-ui/colors'
import createGenerateClassName from 'material-ui/styles/createGenerateClassName'

const theme = createMuiTheme({
  palette: {
    primary: blueGrey,
    secondary: orange,
  },
})

// Configure JSS
const jss = create(jssPreset())
jss.options.createGenerateClassName = createGenerateClassName

export const sheetsManager = new Map()

export default function createContext() {
  return {
    jss,
    theme,
    // This is needed in order to deduplicate the injection of CSS in the page.
    sheetsManager,
    // This is needed in order to inject the critical CSS.
    sheetsRegistry: new SheetsRegistry(),
  }
}
