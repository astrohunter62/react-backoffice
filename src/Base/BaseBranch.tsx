import React from "react";
import classNames from "classnames";
import { withStyles } from "@material-ui/core";
import Drawer from "../Drawer";
import Header from "../Header";
import { MenuDataItem } from "../Menu/Menu";

const drawerWidth = 280;

const styles = (theme: any) => ({
  appFrame: {
    position: "relative",
    display: "flex",
    width: "100%",
    height: "100%",
    minHeight: "100vh",
    transition: "0.25s",
  },
  content: {
    width: `calc(100vw - ${theme.spacing(3) * 2}px)`,
    flexGrow: 1,
    backgroundColor: theme.palette.background.default,
    padding: theme.spacing(3),
    transition: theme.transitions.create("margin", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    height: "calc(100% - 56px)",
    marginLeft: -drawerWidth,
  },
  contentShift: {
    transition: theme.transitions.create("margin", {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
    marginLeft: 0,
  },
});

type BaseBranchProps = {
  open?: boolean;
  title: string;
  menuData: MenuDataItem[];
  isHeaderFixed?: boolean;
  onClick: (...args: any[]) => any;
  handleDrawerOpen: (...args: any[]) => any;
  handleDrawerClose: (...args: any[]) => any;
  redirectTo: (...args: any[]) => any;
  rightContent?: JSX.Element;
  hasHeader: boolean;
  classes: {
    [key: string]: string;
  };
};

const BaseBranch: React.SFC<BaseBranchProps> = ({
  open,
  title,
  menuData,
  isHeaderFixed,
  onClick,
  handleDrawerOpen,
  handleDrawerClose,
  redirectTo,
  rightContent,
  hasHeader = true,
  classes,
  children,
  ...rest
}) => (
  <>
    {hasHeader && (
      <Header
        title={title}
        onDrawerOpen={handleDrawerOpen}
        onClick={onClick}
        isOpen={open}
        isFixed={isHeaderFixed}
      >
        {rightContent || null}
      </Header>
    )}

    <div className={classNames(classes.appFrame)}>
      {hasHeader && (
        <Drawer
          onClose={handleDrawerClose}
          redirectTo={redirectTo}
          isOpen={open}
          data={menuData}
        />
      )}

      <main
        className={classNames(classes.content, {
          [classes.contentShift]: open || !hasHeader,
        })}
      >
        {React.Children.map(children, (child) =>
          React.cloneElement(child as any, {
            ...rest,
          }),
        )}
      </main>
    </div>
  </>
);

BaseBranch.defaultProps = {
  open: false,
  isHeaderFixed: false,
};

export default withStyles(styles as any)(BaseBranch);
