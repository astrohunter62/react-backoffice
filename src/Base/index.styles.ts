import { makeStyles, Theme } from "@material-ui/core";
import { DRAWER_WIDTH } from "../Drawer/Drawer";

const useStyles = makeStyles((theme: Theme) => ({
  "@global html": {
    scrollBehavior: "smooth",
  },

  "@global a, a:link": {
    color: theme.palette.secondary.main,
  },

  appFrame: {
    display: "flex",
  },

  content: {
    width: `calc(100vw - ${theme.spacing(3 * 2)}px)`,
    flexGrow: 1,
    backgroundColor: theme.palette.background.default,
    paddingLeft: theme.spacing(2),
    paddingRight: theme.spacing(2),
    transition: theme.transitions.create("padding", {
      easing: theme.transitions.easing.sharp,
      duration: theme.transitions.duration.leavingScreen,
    }),
    height: "calc(100% - 56px)",
  },

  contentIsOpen: {
    transition: theme.transitions.create("padding", {
      easing: theme.transitions.easing.easeOut,
      duration: theme.transitions.duration.enteringScreen,
    }),
    paddingLeft: DRAWER_WIDTH,
  },

  isHeaderFixed: {
    marginTop: theme.spacing(8),
  },
}));

export default useStyles;
