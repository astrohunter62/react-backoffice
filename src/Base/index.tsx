import React, { FunctionComponent, useState } from "react";
import * as H from "history";
import classNames from "classnames";
import { MenuDataItem } from "../Menu";
import Drawer from "../Drawer";
import Header from "../Header";
import useStyles from "./index.styles";

type BaseProps = {
  title: string;
  menuOpen?: boolean;
  menuData: MenuDataItem[];
  hasHeader?: boolean;
  isHeaderFixed?: boolean;
  history:
    | H.History
    | {
        push: (link: string) => void;
        [key: string]: unknown;
      };
  rightContent?: JSX.Element;
};

const Base: FunctionComponent<BaseProps> = ({
  menuOpen,
  history,
  title,
  menuData,
  isHeaderFixed = false,
  rightContent,
  hasHeader = true,
  children,
  ...rest
}) => {
  const [isOpen, setIsOpen] = useState(Boolean(menuOpen));
  const classes = useStyles();

  const redirectTo = (link: string) => {
    history.push(link);
    window.scrollTo(0, 0);
  };

  return (
    <div className={classNames(classes.appFrame)}>
      {hasHeader && (
        <>
          <Header
            title={title}
            onDrawerOpen={() => setIsOpen(true)}
            onClick={() => redirectTo("/")}
            isOpen={isOpen}
            isFixed={isHeaderFixed}
          >
            {rightContent || null}
          </Header>

          <Drawer
            onClose={() => setIsOpen(false)}
            redirectTo={redirectTo}
            isOpen={isOpen}
            data={menuData}
          />
        </>
      )}

      <main
        className={classNames(classes.content, {
          [classes.contentIsOpen]: isOpen || !hasHeader,
          [classes.isHeaderFixed]: isHeaderFixed,
        })}
      >
        {React.Children.map(children, (child) =>
          React.cloneElement(child as any, {
            ...rest,
          }),
        )}
      </main>
    </div>
  );
};

export default Base;
