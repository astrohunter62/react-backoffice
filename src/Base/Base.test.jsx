import React from 'react'
import MockRouter from 'react-mock-router'
import renderer from 'react-test-renderer'
import Enzyme from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'

import Base from './'

import menuData from '../../__tests__/data/menu'

Enzyme.configure({ adapter: new Adapter() })

describe('Base', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create((
        <MockRouter>
          <Base
            title="Foo"
            menuData={menuData}
            hasCookieInfo
            menuOpen
          >
            <div>Foo</div>
          </Base>
        </MockRouter>
      ))
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly without header', () => {
    const tree = renderer
      .create((
        <MockRouter>
          <Base
            title="Foo"
            menuData={menuData}
            hasCookieInfo
            hasHeader={false}
            menuOpen
          >
            <div>Foo</div>
          </Base>
        </MockRouter>
      ))
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders with drawer open', () => {
    const tree = renderer
      .create((
        <MockRouter>
          <Base
            open
            title="Foo"
            menuData={menuData}
            hasCookieInfo
            hasHeader={false}
            menuOpen
          >
            <div>Foo</div>
          </Base>
        </MockRouter>
      ))
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
