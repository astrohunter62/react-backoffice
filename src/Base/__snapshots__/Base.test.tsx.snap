// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base renders correctly 1`] = `
<Base
  history={
    Object {
      "push": [Function],
    }
  }
  menuData={
    Array [
      Object {
        "title": "Overview",
        "type": "link",
        "url": "/",
      },
      Object {
        "items": Array [
          Object {
            "title": "Dashboard",
            "type": "link",
            "url": "/dashboard",
          },
          Object {
            "title": "List",
            "type": "link",
            "url": "/list",
          },
          Object {
            "title": "Form",
            "type": "link",
            "url": "/form",
          },
          Object {
            "title": "Menu",
            "type": "link",
            "url": "/menu",
          },
          Object {
            "title": "Dialog",
            "type": "link",
            "url": "/dialog",
          },
        ],
        "title": "Components",
        "type": "label",
      },
      Object {
        "isDisabled": true,
        "title": "Disabled Link",
        "type": "link",
        "url": "/portfolio",
      },
      Object {
        "type": "divider",
      },
      Object {
        "icon": <Memo(BugReportIcon) />,
        "title": "Report a bug",
        "type": "link",
        "url": "/bug",
      },
    ]
  }
  menuOpen={true}
  title="Foo"
>
  <div>
    Foo
  </div>
</Base>
`;

exports[`Base renders correctly without header 1`] = `
<Base
  hasHeader={false}
  history={
    Object {
      "push": [Function],
    }
  }
  menuData={
    Array [
      Object {
        "title": "Overview",
        "type": "link",
        "url": "/",
      },
      Object {
        "items": Array [
          Object {
            "title": "Dashboard",
            "type": "link",
            "url": "/dashboard",
          },
          Object {
            "title": "List",
            "type": "link",
            "url": "/list",
          },
          Object {
            "title": "Form",
            "type": "link",
            "url": "/form",
          },
          Object {
            "title": "Menu",
            "type": "link",
            "url": "/menu",
          },
          Object {
            "title": "Dialog",
            "type": "link",
            "url": "/dialog",
          },
        ],
        "title": "Components",
        "type": "label",
      },
      Object {
        "isDisabled": true,
        "title": "Disabled Link",
        "type": "link",
        "url": "/portfolio",
      },
      Object {
        "type": "divider",
      },
      Object {
        "icon": <Memo(BugReportIcon) />,
        "title": "Report a bug",
        "type": "link",
        "url": "/bug",
      },
    ]
  }
  menuOpen={true}
  title="Foo"
>
  <div>
    Foo
  </div>
</Base>
`;

exports[`Base renders with drawer open 1`] = `
<Base
  hasHeader={false}
  history={
    Object {
      "push": [Function],
    }
  }
  menuData={
    Array [
      Object {
        "title": "Overview",
        "type": "link",
        "url": "/",
      },
      Object {
        "items": Array [
          Object {
            "title": "Dashboard",
            "type": "link",
            "url": "/dashboard",
          },
          Object {
            "title": "List",
            "type": "link",
            "url": "/list",
          },
          Object {
            "title": "Form",
            "type": "link",
            "url": "/form",
          },
          Object {
            "title": "Menu",
            "type": "link",
            "url": "/menu",
          },
          Object {
            "title": "Dialog",
            "type": "link",
            "url": "/dialog",
          },
        ],
        "title": "Components",
        "type": "label",
      },
      Object {
        "isDisabled": true,
        "title": "Disabled Link",
        "type": "link",
        "url": "/portfolio",
      },
      Object {
        "type": "divider",
      },
      Object {
        "icon": <Memo(BugReportIcon) />,
        "title": "Report a bug",
        "type": "link",
        "url": "/bug",
      },
    ]
  }
  menuOpen={true}
  title="Foo"
>
  <div>
    Foo
  </div>
</Base>
`;
